### PR TITLE
feat(firmware): let the snap automatically update the firmware

### PR DESCRIPTION
**Pull request by Guillaumebeuzeboc**

### CODE REVIEW

File: turtlebot3c-bringup-snap/snap/hooks/post-refresh [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+
+auto_update=$(snapctl get opencr-auto-update)
+if [ "$auto_update" = "false" ]; then
+  logger -t $SNAP_NAME "OpenCR auto update is disabled. Skipping the update."
+  exit 0
+fi
+
+. "$SNAP/usr/bin/firmware_hash.sh"
+is_hash_the_same
+if [ $? -eq 0 ]; then
```
Comment (artivis) [inline turtlebot3c-bringup-snap/snap/hooks/post-refresh]: Could this be `if is_hash_the_same; then` ?

File: turtlebot3c-bringup-snap/snap/hooks/post-refresh [hunk 2]
```diff
@@ -0,0 +1,26 @@
+#!/bin/sh
+
+logger -t $SNAP_NAME "Post refresh hook starting."
+
+auto_update=$(snapctl get opencr-auto-update)
+if [ "$auto_update" = "false" ]; then
```
Comment (artivis) [inline turtlebot3c-bringup-snap/snap/hooks/post-refresh]: Use curly bracket on bash variables.

File: turtlebot3c-bringup-snap/snap/local/firmware_hash.sh [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+  hash_directory "$SNAP/opencr"
+}
+
+write_firmware_hash() {
+  local firmware_hash=$(hash_firmware_directory)
+  logger -t $SNAP_NAME "OpenCR directory hash is $firmware_hash"
+  echo "${firmware_hash}" > "$SNAP_COMMON/OPENCR_HASH"
+}
+
+is_hash_the_same() {
```
Comment (artivis) [inline turtlebot3c-bringup-snap/snap/local/firmware_hash.sh]:
```suggestion
is_same_hash() {
```

File: turtlebot3c-bringup-snap/snap/local/update_firmware.sh [hunk 4]
```diff
@@ -0,0 +1,26 @@
+#!/bin/sh
+
+arch=$(uname -m)
+
+if [ "$arch" = "x86_64" ]; then
+  echo "System is AMD64, only ARM devices can update the firmware."
```
Comment (artivis) [inline turtlebot3c-bringup-snap/snap/local/update_firmware.sh]:
```suggestion
  echo "System arch is '${arch}'. At the moment only ARM devices can update the firmware."
```

**Review by artivis — CHANGES_REQUESTED** [review]

File: turtlebot3c-bringup-snap/snap/hooks/post-refresh (hunk 2 above)
Comment (Guillaumebeuzeboc) [reply turtlebot3c-bringup-snap/snap/hooks/post-refresh]: applied the fix everywhere.

File: turtlebot3c-bringup-snap/snap/hooks/post-refresh (hunk 1 above)
Comment (Guillaumebeuzeboc) [reply turtlebot3c-bringup-snap/snap/hooks/post-refresh]: Tested and it works. I updated the script.

**Review by artivis — APPROVED** [review]

**Comment by Guillaumebeuzeboc** [comment]
Need a final test to validate the changes.

**Comment by Guillaumebeuzeboc** [comment]
Last commit was tested on robot.

File: turtlebot3c-bringup-snap/snap/local/update_firmware.sh [hunk 5]
```diff
@@ -0,0 +1,26 @@
+#!/bin/sh
+
+arch=$(uname -m)
+
+if [ "${arch}" = "x86_64" ]; then
```
Comment (artivis) [inline turtlebot3c-bringup-snap/snap/local/update_firmware.sh:5]: I understand that this is the one option out of three that's not supported but shouldn't we test for 'not arm' instead? That would be proper.

**Review by artivis — APPROVED** [review]

File: turtlebot3c-bringup-snap/snap/local/update_firmware.sh (hunk 5 above)
Comment (Guillaumebeuzeboc) [reply turtlebot3c-bringup-snap/snap/local/update_firmware.sh:5]: The snaps builds for `arm64`, `armhf` and `x86_64`.  With the `uname` cmd we get the architecture but we don't get the generic ARM or x86. I found it simpler to look for `x86_64` rather looking for the two arm architectures.

**Pull request merged**